go/vt/mysqlctl/compression.go

```diff
@@ -0,0 +1,254 @@
+package mysqlctl
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os/exec"
+	"strings"
+	"sync"
+
+	"github.com/klauspost/compress/zstd"
+	"github.com/klauspost/pgzip"
+	"github.com/pierrec/lz4"
+	"github.com/planetscale/pargzip"
+
+	"vitess.io/vitess/go/vt/logutil"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+var (
+	compressionLevel = flag.Int("compression_level", 1, "What level to pass to the compressor")
+
+	errUnsupportedCompressionEngine    = errors.New("unsupported engine")
+	errUnsupportedCompressionExtension = errors.New("unsupported extension")
+
+	// this is use by getEngineFromExtension() to figure out which engine to use in case the user didn't specify
+	engineExtensions = map[string][]string{
+		".gz":  {"pgzip", "pargzip"},
+		".lz4": {"lz4"},
+		".zst": {"zstd"},
+	}
+)
+
+func getEngineFromExtension(extension string) (string, error) {
+	for ext, eng := range engineExtensions {
+		if ext == extension {
+			return eng[0], nil // we select the first supported engine in auto mode
+		}
+	}
+
+	return "", fmt.Errorf("%w \"%s\"", errUnsupportedCompressionExtension, extension)
+}
+
+func getExtensionFromEngine(engine string) (string, error) {
+	for ext, eng := range engineExtensions {
+		for _, e := range eng {
+			if e == engine {
+				return ext, nil
+			}
+		}
+	}
+
+	return "", fmt.Errorf("%w \"%s\"", errUnsupportedCompressionEngine, engine)
+}
+
+// Validates if the external decompressor exists and return its path.
+func validateExternalCmd(cmd string) (string, error) {
+	if cmd == "" {
+		return "", errors.New("external command is empty")
+	}
+
+	return exec.LookPath(cmd)
+}
+
+func prepareExternalCompressionCmd(ctx context.Context, cmdStr string) (*exec.Cmd, error) {
+	cmdArgs := strings.Split(cmdStr, " ")
```

Comment:
This isn't perfectly cromulent; it will corrupt argument strings with quotes or escaped whitespace. I would suggest bringing https://pkg.go.dev/github.com/google/shlex as a (tiny) dependency to parse the commandline properly.

Reply:
Sounds good, I will take a look at it and make the changes 👍